lib/sprockets/sass/registration.rb

```diff
@@ -0,0 +1,141 @@
+
+      ACTIONS = [
+        { min_comparison: '>=', min: 3, max_comparison: '<', max: 4, action: :register_sprockets_v3 },
+        #    { min_comparison: '>=' , min: 4,  action: :register_sprockets_v4 }
+      ].freeze
+
+      attr_reader :klass, :sprockets_version, :registration_instance, :action_details, :version_selected
+
+      def initialize(klass)
+        @klass = klass
+        @sprockets_version = Sprockets::Sass::Utils.version_of_sprockets
+        @registration_instance = self
+        @action_details = action_to_be_executed
+        @version_selected = @action_details[:min]
+      end
+
+      def run
+        require_libraries
+        send(@action_details[:action])
+      end
+
+      def require_libraries
+        AVAILABLE_VERSIONS.each do |version|
+          require_standard_libraries(version)
+        end
+        require 'sprockets/sass/functions'
+      end
+
+    private
+
+      def require_standard_libraries(version)
+        %w(cache_store compressor functions importer sass_template scss_template).each do |filename|
+          begin
+            require "sprockets/sass/v#{version}/#{filename}"
+          rescue LoadError; end
+        end
+      end
+
+      def register_sprockets_v3_common
+        %w(sass scss).each do |mime|
+          _register_mime_types(mime_type:  "application/#{mime}+ruby", extensions: [".#{mime}.erb", ".css.#{mime}.erb"])
+        end
+        _register_compressors(mime_type: 'text/css', name: :sprockets_sass, klass: Sprockets::Sass::Utils.get_class_by_version('Compressor'))
+      end
+
+      def register_sprockets_v4
+        register_sprockets_v3_common
+        %w(sass scss).each do |mime|
+          _register_transformers(from: "application/#{mime}+ruby", to: "text/#{mime}", klass: Sprockets::ERBProcessor)
+        end
+        _register_v4_preprocessors(
+          Sprockets::Sass::Utils.get_class_by_version('SassTemplate') => ['text/sass', 'text/css'],
+          Sprockets::Sass::Utils.get_class_by_version('ScssTemplate') => ['text/scss', 'text/css']
+        )
+      end
+
+      def register_sprockets_v3
+        register_sprockets_v3_common
+        _register_transformers(
+          { from: 'application/scss+ruby', to: 'text/css', klass: Sprockets::ERBProcessor },
+          from: 'application/sass+ruby', to: 'text/css', klass: Sprockets::ERBProcessor
+        )
+        _register_engines('.sass' => Sprockets::Sass::V3::SassTemplate, '.scss' => Sprockets::Sass::V3::ScssTemplate)
+      end
+
+      def register_sprockets_legacy
+        _register_engines('.sass' => Sprockets::Sass::V2::SassTemplate, '.scss' => Sprockets::Sass::V2::ScssTemplate)
+      end
+
+      def _register_engines(hash)
+        hash.each do |key, value|
+          args = [key, value]
+          args << { mime_type: 'text/css', silence_deprecation: true } if sprockets_version >= 3
+          register_engine(*args)
+        end
+      end
+
+      def action_to_be_executed
```

Comment:
I think it'd be easier to understand what's going on with a simple case_statement. something like:

``` ruby
case @sprockets_version
when 2..3
  register_sprockets_legacy
when 3..4
  register_sprockets_v3
when 4..5
  register_sprockets_v4
else
  raise "Version #{@sprockets_version} is not supported"
end
```